nixos/top-level: improve replaceRuntimeDependencies

alois31 opened the pull request:
## Description of changes

Instead of iterating over all replacements and applying them one by one, use the newly introduced replaceDependencies function to apply them all at once for replaceRuntimeDependencies. The advantages are twofold in case there are multiple replacements:
* Performance is significantly improved, because there is only one pass over the closure to be made.
* Correctness is improved, because replaceDependencies also replaces dependencies of the replacements themselves if applicable.

Further, the option is moved to a `system.replaceDependencies` namespace more consistent with the arguments to the `replaceDependencies` function. This allows wiring up the `cutoffPredicate` functionality, which is also configured to skip the initrd by default because of uselessness in the best case and breakage in the worst case.

Fixes: https://github.com/NixOS/nixpkgs/issues/4336
Fixes: https://github.com/NixOS/nixpkgs/issues/199162

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alois31 committed:
replaceDependencies: evolve from replaceDependency

Rewrite replaceDependency so that it can apply multiple replacements in
one go. This includes correctly handling the case where one of the
replacements itself needs to have another replacement applied as well.
This rewritten function is now aptly called replaceDependencies.

For compatibility, replaceDependency is retained as a simple wrapper
over replaceDependencies. It will cause a rebuild because the unpatched
dependency is now referenced by derivation instead of by storePath, but
the functionality is equivalent.

Fixes: https://github.com/NixOS/nixpkgs/issues/199162

alois31 committed:
nixos/top-level: improve replaceRuntimeDependencies

Instead of iterating over all replacements and applying them one by one,
use the newly introduced replaceDependencies function to apply them all
at once for replaceRuntimeDependencies. The advantages are twofold in
case there are multiple replacements:
* Performance is significantly improved, because there is only one pass
  over the closure to be made.
* Correctness is improved, because replaceDependencies also replaces
  dependencies of the replacements themselves if applicable.

Fixes: https://github.com/NixOS/nixpkgs/issues/4336

alois31 committed:
nixos/top-level: wire up cutoffPackages for replaceDependencies

Move replaceRuntimeDependencies to the replaceDependencies namespace,
where the structure is more consistent with the replaceDependencies
function. This makes space for wiring up cutoffPackages as an option
too.

By default, the system's initrd is excluded. The replacement process does not
work properly anyway due to the structure of the initrd (the files being copied
into it, and it being compressed). In the worst case (which has been observed
to actually occur in practice), a store path makes it into the incompressible
parts of the archive, checksums are broken, and the system won't boot.

roberth commented:
Can't *really* review now, but tests will be needed, and, more superficially, use `let inherit (lib)` or if necessary `let inherit (builtins)` to make the code a little more readable, and a bit faster (don't know if this has hot paths, but it's the second argument for it ;) ).

alois31 commented:
> Can't _really_ review now, but tests will be needed

Between this stuff relying on IFD and Hydra not supporting IFD, I'm stuck in a bit of a bad place here. A test that can't run on Hydra seems pretty pointless, and any ideas I have come up with for circumventing this problem incur a significant cost to other people or infrastructure. If you or anyone else can come up with even a vague idea of how to reasonably test things, I will be glad to implement it.

> and, more superficially, use `let inherit (lib)` or if necessary `let inherit (builtins)` to make the code a little more readable, and a bit faster (don't know if this has hot paths, but it's the second argument for it ;) ).

I was unaware that this style is prefered, and also unaware of the performance benefit (although I doubt it matters that much between creating hundreds of derivations). Changed it.

roberth commented:
You could write a shell script that calls Nix and performs all the assertions, use that for (further) development, and wrap it into a NixOS tests so that Hydra runs it as well.

alois31 commented:
Yeah, that would have been the "significant cost to infrastructure" solution. I'm under the impression that NixOS tests are already on the expensive side, and having to include nixpkgs itself (so that that the configuration with `replaceDependencies` applied can be evaluated) would be particularly bad. If people think this fear is unjustified, I will implement it next week (there should be no need to rush this anyway).

alois31 committed:
replaceDependencies: add tests

The tests cannot be directly built by Hydra, because replaceDependencies relies
on IFD. Instead, they are put inside a NixOS test where they are built on the
guest.

alois31 commented:
Some tests are now implemented (in a separate commit, so that they can be removed easily if they turn out to be too heavy). I also found two bugs (the `replaceDependency` compatibility function not working properly at all, and a precedence issue), which have been fixed in the most recent version.

viperML commented:
Today I just faced with `replaceRuntimeDependencies` requiring IFD.

alois31 commented:
> Today I just faced with `replaceRuntimeDependencies` requiring IFD.

Note that even this improved version requires IFD. Really, that requirement is very unlikely to go away, because in order to know the references, the build result has to be known. (Technically, you could fetch the entire build input closure as a superset of the references (which would be doable in a very hacky way), but that tends to be extremely large.)

alois31 committed:
replaceDirectDependencies: split off from replaceDependencies

This allows both swapping out and reusing the rewrite machinery.

alois31 committed:
replaceDependencies: add support for ca-derivations

Unlike regular input-addressed or fixed-output derivations, floating and
deferred derivations do not have their store path available at evaluation time,
so their outPath is a placeholder. The following changes are needed for
replaceDependencies to continue working:
* Detect the placeholder and retrieve the store path using another IFD hack
  when collecting the rewrite plan.
* Try to obtain the derivation name needed for replaceDirectDependencies from
  the derivation arguments if a placeholder is detected.
* Move the length mismatch detection to build time, since the placeholder has a
  fixed length which is unrelated to the store path.

bryango commented on pkgs/build-support/replace-dependencies.nix:
```diff
@@ -0,0 +1,140 @@
+#
+# In rare cases, it is possible for the replacement process to cause breakage (for example due to checksum mismatch).
+# The cutoffPackages argument can be used to exempt the problematic packages from the replacement process.
+{ drv, replacements, cutoffPackages ? [ ], verbose ? true }:
```
Question: should the `cutoffPackages` default to `map (x: x.newDependency) replacements`, as in the case of the backward-compatible `replaceDependency`?

alois31 replied:
Certainly not, because it completely nullifies one of the two big advantages of `replaceDependencies` (namely, replacing dependencies in other replacements as well); in fact it would even make it worse because the old workaround of choosing the order carefully does not work any more for the NixOS option.

bryango commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1345,7 +1345,18 @@ with pkgs;
+      inherit oldDependency newDependency;
+    }];
+    # When newDependency depends on drv, instead of causing infinite recursion, keep it as is.
+    cutoffPackages = [ newDependency ];
```
As mentioned in the other comment, should the `cutoffPackages` in general default to `map (x: x.newDependency) replacements`?

alois31 replied:
What is a disadvantage in the general case is not applicable here, so keeping backwards compatibility has little cost. If we don't want the difference, then I'd rather remove this here because it only handles a very weird edge case that I assume to be mostly irrelevant in practuce.

bryango reviewed (commented):
Hi @alois31, thank you very much for this wonderful work! I would really love this PR to be merged, and I am slowly reviewing the changes here. This is not easy for me as I am not very familiar with nix internals so it may take a long time... But I will try to post some feedback and questions here as I go through the code. Thank you again for this work!

bryango commented on pkgs/build-support/replace-dependencies.nix:
```diff
@@ -0,0 +1,140 @@
+      echo }) > $out
+    '').outPath;
+
+  targetDerivations = [ drv ]
+    ++ map ({ newDependency, ... }: newDependency) replacements;
```
Why do we need to the original `drv` here? It does not look like a "target" derivation, but rather like a "source" derivation...

alois31 replied:
What I want here (and what the code correctly does) are the (root) derivations that will be needed for building the result, but before the rewrite. This variable was previously called `knownDerivations`, which I think was even worse. I'd be glad if you can come up with an actually good name, because I can't.

dasJ commented on pkgs/build-support/replace-direct-dependencies.nix:
```diff
@@ -0,0 +1,41 @@
+                echo "cannot rewrite $oldBasename to $newBasename: length does not math" >&2
+                exit 1
+            fi
+            echo "s|$oldBasename|$newBasename|g" >> rewrite.sed
```
We could do something like `${oldBasename$/|/\\|}` here to make it safe against names with pipes in them (is that even a thing?)

alois31 replied:
Store path names cannot contain pipes: https://github.com/NixOS/nix/blob/8a8172cd2b5ef2f6dd2d9673a6379447d780ff17/src/libstore/path.cc#L26-L31

alois31 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1345,7 +1345,18 @@ with pkgs;
+    cutoffPackages = [ newDependency ];
+  };
+
+  replaceDirectDependencies = callPackage ../build-support/replace-direct-dependencies.nix { };
```
That's basically the way it was before, see also the sixth commit. Reasons are both that `replaceDirectDependencies` can be useful on its own for some quick hacks without needing IFD, and that it can easily be swapped out with a different implementation. I have made use of both in the past (currently only the latter).

alois31 committed:
replaceDependencies: show warnings as such

alois31 committed:
replaceDependencies: do not build unused replacements

To prevent excessive build times when replacement lists are shared between
partially overlapping closures, skip the build of unused replacements.
Unfortunately, this also means that the replacement won't be applied any more
if another replacement that is applied introduces its source. But this is a
corner case, and we already show a warning, so make it clearer that handling
this situation (should it ever arise) is the responsibility of the user.

RossComputerGuy commented:
Will take a look at this when I have time to.

roconnor commented:
Nice work.
